crates/hash/src/groestl/arch/groestl_multi_avx2.rs

```diff
@@ -0,0 +1,431 @@
+	let mut x = [unsafe { _mm256_set1_epi64x(0) }; 8];
+	let mut y = [unsafe { _mm256_set1_epi64x(0) }; 8];
+	let mut z = [unsafe { _mm256_set1_epi64x(0) }; 8];
+
+	let gf2p8_2: __m256i = unsafe { _mm256_set1_epi8(2) };
+
+	for i in 0..8 {
+		x[i] = unsafe { _mm256_xor_si256(state[i], state[(i + 1) % 8]) };
+	}
+
+	for i in 0..8 {
+		y[i] = unsafe { _mm256_xor_si256(x[i], x[(i + 3) % 8]) };
+	}
+
+	for i in 0..8 {
+		z[i] =
+			unsafe { _mm256_xor_si256(_mm256_xor_si256(x[i], x[(i + 2) % 8]), state[(i + 6) % 8]) };
+	}
+
+	for i in 0..8 {
+		state[i] = unsafe {
+			_mm256_xor_si256(
+				_mm256_gf2p8mul_epi8(
+					gf2p8_2,
+					_mm256_xor_si256(_mm256_gf2p8mul_epi8(gf2p8_2, y[(i + 3) % 8]), z[(i + 7) % 8]),
+				),
+				z[(i + 4) % 8],
+			)
+		};
+	}
+}
+
+fn permutation_p(state: &mut State) {
+	for r in 0..ROUNDS_PER_PERMUTATION {
+		add_round_constant_p(state, r as i8);
+		sub_bytes(state);
+		shift_bytes_p(state);
+		mix_bytes(state);
+	}
+}
+
+fn permutation_q(state: &mut State) {
+	for r in 0..ROUNDS_PER_PERMUTATION {
+		add_round_constant_q(state, r as i8);
+		sub_bytes(state);
+		shift_bytes_q(state);
+		mix_bytes(state);
+	}
+}
+
+#[derive(Clone)]
+pub struct Groestl256Multi {
+	state: State,
+}
+
+impl Groestl256Multi {
+	fn consume_single_block_parallel(&mut self, data: [&[u8]; 4]) {
+		let mut q_data = set_substates_par(data);
+
+		let mut p_data = [unsafe { _mm256_set1_epi64x(0) }; 8];
+
+		for i in 0..8 {
+			p_data[i] = unsafe { _mm256_xor_si256(self.state[i], q_data[i]) };
+		}
+
+		permutation_p(&mut p_data);
+		permutation_q(&mut q_data);
+
+		for i in 0..8 {
+			self.state[i] =
+				unsafe { _mm256_xor_si256(_mm256_xor_si256(self.state[i], q_data[i]), p_data[i]) };
+		}
+	}
+
+	fn finalize(&mut self, out: &mut [MaybeUninit<digest::Output<Groestl256>>; 4]) {
+		let state_copy = self.state;
+		permutation_p(&mut self.state);
+		for (i, state_copy_row) in state_copy.iter().enumerate() {
+			self.state[i] = unsafe { _mm256_xor_si256(self.state[i], *state_copy_row) };
+		}
+
+		let slices = get_substates_par_better(self.state);
+
+		for parallel_idx in 0..NUM_PARALLEL_SUBSTATES {
+			let slice = slices[parallel_idx];
+			unsafe {
+				out[parallel_idx]
+					.as_mut_ptr()
+					.write(*digest::Output::<Groestl256>::from_slice(&slice[32..]));
+			}
+		}
+	}
+}
+
+impl Default for Groestl256Multi {
+	fn default() -> Self {
+		// seeding initial states with the 512b representation of 256
+		Self {
+			state: [
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0) },
+				unsafe { _mm256_set1_epi64x(0x100000000000000u64 as i64) },
+				unsafe { _mm256_set1_epi64x(0) },
+			],
+		}
+	}
+}
+
+impl MultiDigest<4> for Groestl256Multi {
+	type Digest = Groestl256;
+
+	fn new() -> Self {
+		Self::default()
+	}
+
+	fn update(&mut self, data: [&[u8]; 4]) {
```

Comment:
The trait interface allows calling this function several times. How can it work without storing unfinished blocks data in `self`?